operator/k8s_cep_gc.go

```diff
@@ -51,81 +47,48 @@ func enableCiliumEndpointSyncGC() {
-				if err != nil {
-					return err
-				}
-
-				podsCache := map[string]*core_v1.Pod{}
-				for _, pod := range pods.Items {
-					podsCache[pod.Namespace+"/"+pod.Name] = &pod
-				}
-
-				timeNow := time.Now()
-			perCEPFetch:
-				for time.Now().Before(loopStop) { // Guard against no-break bugs
-					time.Sleep(time.Second) // Throttle lookups in case of a busy loop
-
-					ceps, err := ciliumClient.CiliumEndpoints(meta_v1.NamespaceAll).List(ctx, listOpts)
-					switch {
-					case err != nil && k8serrors.IsResourceExpired(err) && ceps.Continue != "":
-						// This combination means we saw a 410 ResourceExpired error but we
-						// can iterate on the now-current snapshot. We need to refetch,
-						// however.
-						// See https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L350-L381
-						// or the docs for k8s.io/apimachinery/pkg/apis/meta/v1.ListOptions
-						// vendored into this repo.
-						listOpts.Continue = ceps.Continue
-						continue perCEPFetch
-
-					case err != nil:
-						scopedLog.WithError(err).Debug("Cannot list CEPs")
-						return err
+				// For each CEP we fetched, check if we know about it
+				for _, cepObj := range watchers.CiliumEndpointStore.List() {
+					cep, ok := cepObj.(*cilium_v2.CiliumEndpoint)
+					if !ok {
+						log.WithField(logfields.Object, cepObj).
+							Errorf("Saw %T object while expecting *cilium_v2.CiliumEndpoint", cepObj)
 					}
 
-					// setup listOpts for the next iteration
-					listOpts.Continue = ceps.Continue
-
-					// For each CEP we fetched, check if we know about it
-					for _, cep := range ceps.Items {
-						if cep.Status.Identity != nil {
-							identityHeartbeat.MarkAlive(strconv.FormatInt(cep.Status.Identity.ID, 10), timeNow)
-						}
-
-						cepFullName := cep.Namespace + "/" + cep.Name
-						_, exists := podsCache[cepFullName]
-						if !exists {
-							// delete
-							scopedLog = scopedLog.WithFields(logrus.Fields{
-								logfields.EndpointID: cep.Status.ID,
-								logfields.K8sPodName: cepFullName,
-							})
-							scopedLog.Debug("Orphaned CiliumEndpoint is being garbage collected")
-							PropagationPolicy := meta_v1.DeletePropagationBackground // because these are const strings but the API wants pointers
-							if err := ciliumClient.CiliumEndpoints(cep.Namespace).Delete(
-								ctx,
-								cep.Name,
-								meta_v1.DeleteOptions{PropagationPolicy: &PropagationPolicy}); err != nil {
-								scopedLog.WithError(err).Debug("Unable to delete orphaned CEP")
-								return err
-							}
-						}
+					cepFullName := cep.Namespace + "/" + cep.Name
+					_, exists, err := watchers.PodStore.GetByKey(cepFullName)
+					if err != nil {
+						scopedLog.WithError(err).Warn("Unable to get pod from store")
+						continue
 					}
-					if ceps.Continue != "" {
-						// there is more data, continue
-						continue perCEPFetch
+					if !exists {
+						// FIXME: this is fragile has we might have received the
```

Comment:
```suggestion
						// FIXME: this is fragile as we might have received the
```
is this what you meant?